Bottom: 633f59bb43f60ff32cfbccaf47ee0d6d478d9802
Top:    b6318125181ccd3e12bc852d5c8f1ff0d9ac2a76
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-08-07 14:06:31 +0800

macvlan: XDP rx handler support

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/macvlan.c b/drivers/net/macvlan.c
index f78319f..d2ef1aa 100644
--- a/drivers/net/macvlan.c
+++ b/drivers/net/macvlan.c
@@ -436,6 +436,35 @@ static void macvlan_forward_source(struct sk_buff *skb,
 	}
 }
 
+/* called under rcu_read_lock() from XDP handler */
+static rx_handler_result_t macvlan_handle_xdp(struct xdp_buff *xdp)
+{
+	const struct ethhdr *eth = (const struct ethhdr *)xdp->data;
+	struct xdp_rxq_info *xdp_rxq = xdp->rxq;
+	struct net_device *dev;
+	struct macvlan_port *port;
+	struct macvlan_dev *vlan;
+
+	if (is_multicast_ether_addr(eth->h_dest))
+		return RX_HANDLER_PASS;
+
+	port = macvlan_port_get_rcu(xdp_rxq->dev);
+	if (port->source_count)
+		return RX_HANDLER_PASS;
+
+	if (macvlan_passthru(port))
+		vlan = list_first_or_null_rcu(&port->vlans,
+					      struct macvlan_dev, list);
+	else
+		vlan = macvlan_hash_lookup(port, eth->h_dest);
+
+	dev = vlan->dev;
+	if (unlikely(!(dev->flags & IFF_UP)))
+		return RX_HANDLER_PASS;
+
+	return RX_HANDLER_CONSUMED;
+}
+
 /* called under rcu_read_lock() from netif_receive_skb */
 static rx_handler_result_t macvlan_handle_frame(struct sk_buff **pskb)
 {
@@ -1172,7 +1201,8 @@ static int macvlan_port_create(struct net_device *dev)
 	skb_queue_head_init(&port->bc_queue);
 	INIT_WORK(&port->bc_work, macvlan_process_broadcast);
 
-	err = netdev_rx_handler_register(dev, macvlan_handle_frame, port);
+	err = netdev_rx_handler_register_xdp(dev, macvlan_handle_frame,
+					     macvlan_handle_xdp, port);
 
 	if (err)
 		kfree(port);
